b2/thread_pool.py

```diff
@@ -100,6 +103,21 @@ def get_exceptions(self):
         with self._lock:
             return self._exceptions
 
+    def shut_down(self):
+        # Nobody can add any tasks any more.
+        with self._lock:
+            queue = self._queue
+            self._queue = None
```

Comment:
if `shut_down` will race with `add_task` or `join`, 'None has no attribute 'join' will be the error we get. I suggest using a dummy queue object instead to raise an exception ("unable to add tasks to a queue during shutdown" or something)